dumping entire MapHeaderPtrs table and associated data structures

mdraves91 opened the pull request:
This PR dumps the entire MapHeaderPtrs table and associated data structures. I wrote a new script, `map_header_extractor.py`, based on the skeleton of `script_extractor.py`, to dump the table. This is the output directly from the script with no edits. Function pointers in data structures are left as raw bytes, instead of creating labels and disassembling every function.

You can run the script yourself with `python tools/map_header_extractor.py > out.asm`

### Some notes:
1)
It is a bit tricky to reach some many of the data structures, because they are not pointed to directly by `xxx_MapHeader`, but they are loaded by functions pointed to in `xxx_MapScripts`. For these, there is a consistent `call Func_xxx` instruction AFTER the instruction that loads the data table we want, so you will see some gnarly if statements looking in the ROM for the bytes corresponding to that call, and a comment explaining which function we are looking for.

An example of one of the these functions we have to search for a data table:
```
Func_2c0c8:
	ld hl, IshiharasHouse_NPCs
	call Func_3205
	scf
	ccf
	ret
```

If you do run the script yourself, you will see 62 warning messages about tables not being found using this method.

```
WARN: MasonLaboratorySide2_OWInteractions table not found in function 3c825
```
I have checked all of them by hand, and those data tables indeed do not exist. There used to be 3 more warned as missing that *did* in fact exist, but I added special handling in the script to catch them.

2)
There are a handful of step functions that do not fit neatly into the new `map_exit` or `ow_script` macros, so I had the script dump them using the full `_ow_coordinate_function` macro. They are rare enough that I don't think it's worth trying to fully understand them and create new macros for them at the moment.

3)
I left in the `; gap from 0x3cf98 to 0x3cfd3` comments, so we will know where to disassemble next.

4)
like my previous PR, I left the many, many functions pointed to by these data tables as raw bytes. I did add functionality to the script to dump them as FUNC_xxx labels instead (use the -f flag), which will come in handy when it's time to disassemble them.

mdraves91 commented:
Phase 2 will be to try and extend the script to dump all of the functions in these data structures too, because there are over 1200 of them. It is easy enough to just call `tcg2disasm.py` on every function, but there are a *lot* that have baked in event scripts which causes problems for the disassembler.

I came across this particularly gnarly example that I had to disassemble partially by hand, because it does a `script_command_08 ` jump to a totally disconnected chunk of event script, and there is other assembly code before and after both of these script blocks.

```
Func_3468f:
	ld a, $32
	ld [$d60e], a
	ld hl, $a2c
	ld a, l
	ld [wd60f], a
	ld a, h
	ld [$d610], a
	xor a
	call Func_33f2
Script_346a3:
		script_command_01
		script_command_10 $0b
		script_command_0a .ows_346c9
		script_command_10 $aa
		script_command_0a .ows_346b6
		script_command_0e $aa
		script_command_05 Text081f
		script_command_08 .ows_346b9
.ows_346b6
		script_command_05 Text0820
.ows_346b9
		script_command_02
		end_script
	ret
	ld a, $d9
	farcall GetEventValue
	jr z, .asm_346c6
	scf
	ret
.asm_346c6
	scf
	ccf
	ret
.ows_346c9
		script_command_05 Text0821
		script_command_02
		script_command_2d $46f0
		script_command_2b $46f3, $01
		script_command_2f
		end_script
	ld a, [wPlayerOWObject]
	farcall Func_10da7
	ld c, e
	ld a, $3e
	farcall SetVarValue
	ld a, $1f
	ld de, $1f1f
	ld b, $01
	farcall Func_d3c4
	ret
; 0x346f0
```

dannye reviewed (approved):
Wow! You made huge progress on this very quickly, thank you very much. Looks great.

dannye commented:
In commit e78fd39 I made some minor cleanup, hope you don't mind. I added `GameCenter1_NPCInteractions` and `IshiharasVilla1_NPCInteractions`, which are being dumped by your script but weren't included in this PR for some reason. Let me know if that's wrong.

I also removed the imports between the two extractor scripts. Since it was only importing 3 very tiny functions, might as well make the two scripts entirely independent. (It was also importing `sort_and_filter` which is not so tiny, but that function was being overridden anyway.)
And I cleaned up the arg parser options.

mdraves91 commented:
> In commit [e78fd39](https://github.com/pret/poketcg2/commit/e78fd395f3352c9230dbfe9f6cb52f69c6c68a40) I made some minor cleanup, hope you don't mind. I added `GameCenter1_NPCInteractions` and `IshiharasVilla1_NPCInteractions`, which are being dumped by your script but weren't included in this PR for some reason. Let me know if that's wrong.
Thanks for catching this. Not sure how I missed them in the PR... I think the special handling I added in [9c851d7](https://github.com/pret/poketcg2/pull/31/commits/9c851d75c8a3f01cef40c0b4ca4f1be4fa76b3eb) for 3 *other* maps also caught these two missing ones, and I forgot to copy them over.

> I also removed the imports between the two extractor scripts. Since it was only importing 3 very tiny functions, might as well make the two scripts entirely independent. (It was also importing `sort_and_filter` which is not so tiny, but that function was being overridden anyway.) And I cleaned up the arg parser options.

No problem. I originally intended to import a lot more functions, but then found that a lot of the ones I wanted to import from `script_extractor.py` were using some global variables that I didn't want to refactor.

dannye commented:
> Phase 2 will be to try and extend the script to dump all of the functions in these data structures too, because there are over 1200 of them. It is easy enough to just call `tcg2disasm.py` on every function, but there are a _lot_ that have baked in event scripts which causes problems for the disassembler.
> 
> I came across this particularly gnarly example that I had to disassemble partially by hand, because it does a `script_command_08 ` jump to a totally disconnected chunk of event script, and there is other assembly code before and after both of these script blocks.

(This isn't necessarily helping to solve the problem, but just sharing some thoughts..)

For TCG 1, what I did for the overworld scripts was to simply collect a giant list of addresses of _all known ow scripts_ and keep that list in my notes.

Then, every time we updated a script command macro name or argument list, I would re-run script_extractor2.py with all the known addresses at one time (since the python script is configured to take an unlimited number of address arguments), like:
```sh
python tools/script_extractor2.py <*giant* list of addresses> | clip
```
Then I would paste the clipboard into a blank file and then cut/paste the portions wherever they need to go.

This is obviously not perfect and still a little tedious. Especially collecting the list of known addresses, but at least that only needs to be done one time.

This was doable for TCG 1 because all the scripts are in a single bank, and of course TCG 1 is a much smaller game. But this still might be how I approach dumping the scripts for TCG 2.

dannye commented:
> No problem. I originally intended to import a lot more functions, but then found that a lot of the ones I wanted to import from `script_extractor.py` were using some global variables that I didn't want to refactor.

Yep, I figured that's exactly how it happened, haha.

Just like you made `map_header_extractor.py` by copying and modifying `script_extractor.py`, my `script_extractor.py` is based on my own [crystal_audio.py](https://github.com/pret/pokemon-reverse-engineering-tools/blob/master/pokemontools/crystal_audio.py) for dumping all song files in Crystal.
